akka-projection-jdbc/src/main/scala/akka/projection/jdbc/internal/JdbcOffsetStore.scala

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.projection.jdbc.internal
+
+import java.sql.Connection
+import java.sql.Timestamp
+import java.time.Clock
+import java.time.Instant
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+import akka.Done
+import akka.annotation.InternalApi
+import akka.projection.MergeableOffset
+import akka.projection.ProjectionId
+import akka.projection.internal.OffsetSerialization
+import akka.projection.internal.OffsetSerialization.SingleOffset
+import akka.projection.internal.OffsetSerialization.toStorageRepresentation
+import akka.projection.internal.OffsetSerialization.fromStorageRepresentation
+import akka.projection.jdbc.javadsl.JdbcSession
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] class JdbcOffsetStore(settings: JdbcSettings, clock: Clock) {
+
+  def createIfNotExists[S <: JdbcSession](jdbcSessionFactory: () => S)(implicit ec: ExecutionContext): Future[Done] = {
+    JdbcSession
+      .runBlockingDBIO(jdbcSessionFactory) { sess =>
+        sess.withConnection { conn =>
+          val stmt = conn.createStatement()
+          stmt.execute(settings.dialect.createTableStatement)
+          stmt.execute(settings.dialect.alterTableStatement)
+          Done
+        }
+      }
+  }
+
+  def clearOffset(conn: Connection, projectionId: ProjectionId) = {
+    val stmt = conn.prepareStatement(settings.dialect.clearOffsetStatement)
+    stmt.setString(1, projectionId.name)
+    stmt.setString(2, projectionId.key)
+    stmt.executeUpdate()
+  }
+
+  def readOffset[Offset](conn: Connection, projectionId: ProjectionId): Option[Offset] = {
+    val stmt = conn.prepareStatement(settings.dialect.readOffsetQuery)
+    stmt.setString(1, projectionId.name)
+
+    val resultSet = stmt.executeQuery()
+
+    // FIXME: need to iterate over all results to build MultipleOffsets
+    if (resultSet.first()) {
+      val offsetStr = resultSet.getString(3)
+      val manifest = resultSet.getString(4)
```

Comment:
could use column names instead, or is it an advantage to use index?

Reply:
Yes, we can. I don't think there is an advantage of using index.

Reply:
I didn't check the real code, but it probably has a reversed index. So, there is a string comparison and then lookup on the index by int. Anyway, negligible for this kind of operation. 

i changed to make it more obvious.